server/api/sample_sizes.py

```diff
@@ -186,18 +202,72 @@ def sample_sizes_for_contest(contest: Contest):
     }
 
 
+@background_task
+def first_round_sample_size_options(election_id: str):
+    election = Election.query.get(election_id)
+    election.sample_size_options = sample_size_options(election, round_one=True)
+
+
+def serialize_sample_size_options(sample_size_options):
+    if sample_size_options is None:
+        return None
+    return {
+        contest_id: list(options.values())
+        for contest_id, options in sample_size_options.items()
+    }
+
+
 @api.route("/election/<election_id>/sample-sizes", methods=["GET"])
 @restrict_access([UserType.AUDIT_ADMIN])
 def get_sample_sizes(election: Election):
-    sample_sizes = {
-        contest_id: list(options.values())
-        for contest_id, options in sample_size_options(election, round_one=True).items()
-    }
+    task = serialize_background_task(election.sample_size_options_task)
+
     # If we've already started the first round, return which sample size was
     # selected for each contest so we can show the user
-    selected_sample_sizes = dict(
-        RoundContest.query.join(Round)
-        .filter_by(election_id=election.id, round_num=1)
-        .values(RoundContest.contest_id, RoundContest.sample_size)
+    if len(list(election.rounds)) > 0:
+        selected_sample_sizes = dict(
+            RoundContest.query.join(Round)
+            .filter_by(election_id=election.id, round_num=1)
+            .values(RoundContest.contest_id, RoundContest.sample_size)
+        )
+        return jsonify(
+            sampleSizes=serialize_sample_size_options(election.sample_size_options),
+            selected=selected_sample_sizes,
+            task=task,
+        )
+
+    # If we're still in audit setup, and we recently computed sample sizes, return them.
+    if (
+        election.sample_size_options_task
+        and election.sample_size_options_task.completed_at
+    ):
+        age = (
+            datetime.now(timezone.utc) - election.sample_size_options_task.completed_at
+        )
+        if age < timedelta(seconds=5):
+            return jsonify(
+                sampleSizes=serialize_sample_size_options(election.sample_size_options),
+                selected=None,
+                task=task,
+            )
+
+    # Otherwise, start a background task to compute sample size options (as
+    # long as there isn't already one in progress).
+    if task is None or (
+        task["status"]
+        not in [ProcessingStatus.READY_TO_PROCESS, ProcessingStatus.PROCESSING]
+    ):
+        election.sample_size_options = None
+        election.sample_size_options_task = create_background_task(
+            first_round_sample_size_options, dict(election_id=election.id)
+        )
+        db_session.commit()
+
+    # In tests, the background task will complete immediately, so we return
```

Comment:
this makes me wonder if there's anything we can do, in tests, to make background tasks not happen immediately, to queue them up on some event loop... I know no such thing, but you know what i mean.

for later though.

Reply:
there's a config variable for whether or not to run the tests immediately. i designed it this way. i explicitly disable that behavior in one of the tests i added that actually checks all the various states of the background task. but for most tests, it's simpler to just have it complete immediately.